### PR TITLE
update release-1.24 to go1.19

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Picks commits needed to update release-1.24 to go1.19, since go1.18 goes out of support in ~Feb 2023. go1.19 is supported through ~August 2023, which covers the remainder of the [1.24 releases](https://kubernetes.io/releases/patch-releases/#1-24).

This ensures that release-1.24 can pick up future runtime security fixes.

This happens to be possible because the only behavior change we needed to adapt to in go1.19 was the ErrDot LookPath change, which we were able to accommodate in-process when release-1.25 updated to go1.19.

This PR includes commits from master branch go1.19 update:
* https://github.com/kubernetes/kubernetes/pull/110811 (needed to evaluate stdlib 1.19 packages with generics)
* https://github.com/kubernetes/kubernetes/pull/111254
  * includes `Adjust for os/exec changes in 1.19` commit, which adjusts to preserve existing behavior with the go1.19 change to error when locating binaries in the current directory
  * also includes 1.19 gofmt rewritten comments, which makes the overall diff look way bigger without causing any functional changes
* https://github.com/kubernetes/kubernetes/pull/111696

Before this can merge, 1.24 CI would need to be updated to use go1.19, so that gofmt and dependency CI checks pass (xref https://github.com/kubernetes/release/issues/2822)

#### Which issue(s) this PR fixes:

xref https://github.com/kubernetes/kubernetes/issues/112408

#### Does this PR introduce a user-facing change?
```release-note
Kubernetes 1.24 is now built with go1.19.4
```

### CODE REVIEW

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by dims** [comment]
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

File: build/build-image/cross/VERSION
Comment (cpanato) [inline build/build-image/cross/VERSION:1]: you can replace with `v1.24.0-go1.19.4-bullseye.0` now :)

Comment (liggitt) [reply build/build-image/cross/VERSION]: done

**Comment by liggitt** [comment]
- [x] pull-kubernetes-dependencies-go-canary - https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/113956/pull-kubernetes-dependencies-go-canary/1603876137908310016
- [ ] pull-kubernetes-integration-go-canary
- [x] pull-kubernetes-unit-go-canary - https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/113956/pull-kubernetes-unit-go-canary/1603876138013167616
- [ ] pull-kubernetes-verify-go-canary

**Comment by liggitt** [comment]
/cc @dims

This is ready for review, all required presubmits are green (or are green on the go-canary variants that use go1.19, which `verify` and `dependencies` need)

File: staging/src/k8s.io/client-go/tools/watch/until.go
Comment (aojea) [inline staging/src/k8s.io/client-go/tools/watch/until.go:104]: lots of identantion changes, is that normal?

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/watch/until.go:104]: Yeah, go1.19 gofmt changed dramatically :-(

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/watch/until.go:104]: xref https://tip.golang.org/doc/go1.19#go-doc

Comment (aojea) [reply staging/src/k8s.io/client-go/tools/watch/until.go:104]: TIL

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/watch/until.go:104]: Still an open issue against master to resolve incorrect changes gofmt made to comments - https://github.com/kubernetes/kubernetes/issues/111701

**Comment by dims** [comment]
/approve
/lgtm

Looks like we still have a couple of unhappy CI jobs. the rest of the commits in this PR look good. thanks for separating them into logical commits. Also separating the mechanical ones was a good idea as well.

**Comment by liggitt** [comment]
> Looks like we still have a couple of unhappy CI jobs


That's the `verify` job (failing because it doesn't like 1.19 gofmt) and `dependencies` job (failing because it doesn't like 1.19 go.mod files). The go-canary versions of those jobs running on go1.19 have already passed, so the next step is to get CI for release-1.24 shifted to use go1.19.4, then presubmits here will pass

**Comment by liggitt** [comment]
/assign @dchen1107 @derekwaynecarr @mrunalp 
needs cri-api approver since this touches a go.mod file and gofmt reformatted comments within that module (no other changes)

File: hack/lib/golang.sh
Comment (aojea) [inline hack/lib/golang.sh:483]: needs to switch?

Comment (liggitt) [reply hack/lib/golang.sh:483]: yeah, I'll update this once we switch presubmit images to run with go1.19, it would have turned all CI jobs red before that

**Comment by dchen1107** [comment]
/approve from SIG Node.

Thanks for taking care of this for the community.

File: hack/lib/golang.sh
Comment (liggitt) [reply hack/lib/golang.sh:483]: Actually, I think we should avoid bumping this until we're happy with post-submits and maybe until we cut the next 1.24 patch release. As soon as we do this, anything currently happily building off of release-1.24 with go1.18 will start failing, and I don't want to ripple out build/config changes past the immediate k8s.io/test-infra and k8s.io/release updates we're making quite yet

**Comment by liggitt** [comment]
https://github.com/kubernetes/test-infra/pull/28290 updated kubekins and krte images for 1.24

/test all

**Comment by liggitt** [comment]
Previously failing `pull-kubernetes-dependencies` now passes on go1.19.4 🎉

**Comment by liggitt** [comment]
all presubmits green, has lgtm + approve

cc @kubernetes/release-managers

File: build/dependencies.yaml
Comment (xmudrii) [inline build/dependencies.yaml:105]: This should be `1.<major>`, i.e. `1.19`. See on master for example: https://github.com/kubernetes/kubernetes/blob/7814d10882d5f159630ccbbaf17c3e014d4f7835/build/dependencies.yaml#L105

Comment (xmudrii) [reply build/dependencies.yaml:105]: As discussed on Slack, it will be fixed in a follow-up PR along with the comment from @aojea.

Ref: https://kubernetes.slack.com/archives/C2C40FMNF/p1671026815368429

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve

File: hack/lib/golang.sh
Comment (liggitt) [reply hack/lib/golang.sh:483]: opened https://github.com/kubernetes/kubernetes/pull/114616 with this follow-up

File: build/dependencies.yaml
Comment (liggitt) [reply build/dependencies.yaml:105]: opened https://github.com/kubernetes/kubernetes/pull/114616 with this follow-up

**Comment by liggitt** [comment]
tracking status of postsubmits after this merged (commit 9e7bdc512)

https://testgrid.k8s.io/sig-release-1.24-blocking
* ✅ 3 green runs on https://testgrid.k8s.io/sig-release-1.24-blocking#build-1.24
* ✅ 1 green run on https://testgrid.k8s.io/sig-release-1.24-blocking#gce-cos-k8sstable2-alphafeatures
* ✅ 1 green run on https://testgrid.k8s.io/sig-release-1.24-blocking#gce-cos-k8sstable2-default
* ✅ 1 green run on https://testgrid.k8s.io/sig-release-1.24-blocking#gce-cos-k8sstable2-ingress
* ✅ 1 green run on https://testgrid.k8s.io/sig-release-1.24-blocking#gce-cos-k8sstable2-reboot
* __ runs on https://testgrid.k8s.io/sig-release-1.24-blocking#gce-device-plugin-gpu-1.24
* __ runs on https://testgrid.k8s.io/sig-release-1.24-blocking#gce-cos-1.24-scalability-100
* ⚠️ 1 green run with failed `k8s.io/kubernetes/test/integration/scheduler` tests on https://testgrid.k8s.io/sig-release-1.24-blocking#integration-1.24 (`k8s.io/kubernetes/test/integration/scheduler` tests have been _very_ unstable since at least 2022-12-07 on that job even prior to this merge - see [history](https://testgrid.k8s.io/sig-release-1.24-blocking#integration-1.24&width=20))
* ✅ 1 green run on https://testgrid.k8s.io/sig-release-1.24-blocking#kind-1.24-parallel
* ✅ 1 green run on https://testgrid.k8s.io/sig-release-1.24-blocking#kind-ipv6-1.24-parallel
* ✅ 3 green runs on https://testgrid.k8s.io/sig-release-1.24-blocking#ci-kubernetes-unit-1-24
* ✅ 1 green run on https://testgrid.k8s.io/sig-release-1.24-blocking#verify-1.24

https://testgrid.k8s.io/sig-release-1.24-informing
* __ runs on https://testgrid.k8s.io/sig-release-1.24-informing#capz-windows-containerd-1.24 (flaking before this merge)
* __ runs on https://testgrid.k8s.io/sig-release-1.24-informing#gce-cos-k8sstable2-serial (flaking before this merge)
* ✅ 1 green run on https://testgrid.k8s.io/sig-release-1.24-informing#gce-cos-k8sstable2-slow (flaking before this merge)
* __ runs on https://testgrid.k8s.io/sig-release-1.24-informing#kubeadm-kinder-1-24
* __ runs on https://testgrid.k8s.io/sig-release-1.24-informing#kubeadm-kinder-1-24-on-1-23
* __ runs on https://testgrid.k8s.io/sig-release-1.24-informing#kubeadm-kinder-upgrade-1-23-1-24

**Pull request merged**
